lib/internal/streams/readable.js

```diff
@@ -1284,9 +1322,265 @@ ObjectDefineProperties(ReadableState.prototype, {
+  bodyUsed: {
+    get() {
+      return isDisturbed(this);
+    }
+  },
+
+  body: {
+    get() {
+      if (this[kConsume]?.type === kWebStreamType) {
+        return this[kConsume].stream;
+      }
+
+      return consume(this, kWebStreamType);
+    }
+  },
+
+  text: {
+    get() {
+      return consume(this, kTextType);
+    }
+  },
+
+  json: {
+    get() {
+      return consume(this, kJSONType);
+    }
+  },
+
+  blob: {
+    get() {
+      return consume(this, kBlobType);
+    }
+  },
+
+  arrayBuffer: {
+    get() {
+      return consume(this, kArrayBufferType);
+    }
   }
 });
 
+function isLocked(self) {
+  return self[kConsume]?.stream?.locked === true;
+}
+
+// https://streams.spec.whatwg.org/#readablestream-disturbed
+function isDisturbed(self) {
+  return self.destroyed || self.readableDidRead;
+}
+
+// https://fetch.spec.whatwg.org/#body-unusable
+function isUnusable(self) {
+  return isDisturbed(self) || isLocked(self);
+}
+
+function consume(self, type) {
+  if (isUnusable(self)) {
+    // eslint-disable-next-line no-restricted-syntax
+    throw new TypeError('unusable');
+  }
+
+  if (type === kWebStreamType) {
+    if (!ReadableStream) {
+      ReadableStream = require('internal/webstreams/readablestream')
+        .ReadableStream;
+    }
+
+    const objectMode = self.readableObjectMode;
+    const highWaterMark = self.readableHighWaterMark;
+    // When not running in objectMode explicitly, we just fall
+    // back to a minimal strategy that just specifies the highWaterMark
+    // and no size algorithm. Using a ByteLengthQueuingStrategy here
+    // is unnecessary.
+    let strategy;
+    if (objectMode) {
+      if (!CountQueuingStrategy) {
+        CountQueuingStrategy = require('internal/webstreams/queuingstrategies');
+      }
+      strategy = new CountQueuingStrategy({ highWaterMark });
+    } else {
+      strategy = { highWaterMark };
+    }
+
+    self
+      .on('error', function(err) {
+        const { controller } = this[kConsume];
+        controller.error(err);
+      })
+      .on('close', function() {
+        const { controller } = this[kConsume];
+        if (controller) {
+          controller.error(new AbortError());
+        }
+      });
+    const consume = self[kConsume] = {
+      type,
+      objectMode,
+      controller: null,
+      push(chunk) {
+        const { objectMode, controller } = this;
+
+        assert(controller);
+
+        if (chunk === null) {
+          // TODO: Do we need to wait for construct to finish?
+          controller.close();
+          this.controller = null;
+        } else {
+          if (!objectMode) {
+            if (typeof chunk === 'string') {
+              chunk = new Uint8Array(Buffer.from(chunk));
+            } else if (Buffer.isBuffer(chunk)) {
+              // Copy the Buffer to detach it from the pool.
+              chunk = new Uint8Array(chunk);
+            } else if (Stream._isUint8Array(chunk)) {
```

Comment:
Wouldn't this be the most common case? Also aren't buffers Uint8Arrays anyway?

Reply:
> Wouldn't this be the most common case?

I don't think so. `Buffer` is more common.

> Also aren't buffers Uint8Arrays anyway?

Yes, but  `Buffer` overrides `Uint8Array` methods providing different behaviour.